src/storage/sftp/write.c

```diff
@@ -0,0 +1,461 @@
+
+    do
+    {
+        rc = libssh2_sftp_unlink_ex(this->sftpSession, strZ(this->interface.name), (unsigned int)strSize(this->interface.name));
+    }
+    while (rc == LIBSSH2_ERROR_EAGAIN && waitMore(wait));
+
+    if (rc)
+    {
+        storageSftpEvalLibSsh2Error(
+            rc, libssh2_sftp_last_error(this->sftpSession), &FileRemoveError,
+            strNewFmt("unable to remove existing '%s'", strZ(this->interface.name)), NULL);
+    }
+
+    FUNCTION_LOG_RETURN_VOID();
+}
+
+/***********************************************************************************************************************************
+Rename a file
+***********************************************************************************************************************************/
+static void
+storageWriteSftpRename(THIS_VOID)
+{
+    THIS(StorageWriteSftp);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(STORAGE_WRITE_SFTP, this);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+
+    int rc = 0;
+    Wait *wait = waitNew(this->timeoutConnect);
+
+    do
+    {
+        rc = libssh2_sftp_rename_ex(
+            this->sftpSession, strZ(this->nameTmp), (unsigned int)strSize(this->nameTmp), strZ(this->interface.name),
+            (unsigned int)strSize(this->interface.name),
+            LIBSSH2_SFTP_RENAME_OVERWRITE | LIBSSH2_SFTP_RENAME_ATOMIC | LIBSSH2_SFTP_RENAME_NATIVE);
+    }
+    while (rc == LIBSSH2_ERROR_EAGAIN && waitMore(wait));
+
+    if (rc)
+    {
+        storageSftpEvalLibSsh2Error(
+            rc, libssh2_sftp_last_error(this->sftpSession), &FileRemoveError,
+            strNewFmt("unable to move '%s' to '%s'", strZ(this->nameTmp), strZ(this->interface.name)), NULL);
+    }
+
+    FUNCTION_LOG_RETURN_VOID();
+}
+
+/***********************************************************************************************************************************
+Close the file
+***********************************************************************************************************************************/
+static void
+storageWriteSftpClose(THIS_VOID)
+{
+    THIS(StorageWriteSftp);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(STORAGE_WRITE_SFTP, this);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+
+    Wait *wait = NULL;
+
+    // Close if the file has not already been closed
+    if (this->sftpHandle != NULL)
+    {
+        int rc;
+        char *libSsh2ErrMsg;
+        int errMsgLen;
+        int libSsh2ErrNo;
+
+        if (this->interface.syncFile)
+        {
+            wait = waitNew(this->timeoutConnect);
+
+            do
+            {
+                rc = libssh2_sftp_fsync(this->sftpHandle);
+            }
+            while (rc == LIBSSH2_ERROR_EAGAIN && waitMore(wait));
+
+            if (rc)
+                THROW_FMT(FileSyncError, STORAGE_ERROR_WRITE_SYNC, strZ(this->nameTmp));
+        }
+
+        // Update modified time
+        if (this->interface.timeModified != 0)
```

Comment:
This code is unreachable outside of unit tests. Backups/archive-push will never pass a modified timestamp to storageNewWriteP().

Reply:
removed unreachable code, commit f25431f5b67377280f90996e8aead476d88c2b6a